src/main/python/deep_qa/layers/attention/gated_attention.py

```diff
@@ -0,0 +1,99 @@
+
+        - 1. :math:`\alpha_i = softmax(Q^T d_i)`
+        - 2. :math:`q_i = Q \alpha_i`
+        - 3. :math:`x_i = d_i \odot q_i` (:math:`\odot` is element-wise multiplication)
+
+    This layer implements equations 2 and 3 above but in a batched manner to get
+    :math:`X`, a tensor with all :math:`x_i`. Thus, the input
+    to the layer is :math:`\alpha` (``normalized_qd_attention``), a tensor with
+    all :math:`\alpha_i`, as well as :math:`Q` (``question_matrix``), and
+    :math:`D` (``document_matrix``), a tensor with all :math:`d_i`. Equation 6 uses
+    element-wise multiplication to model the interactions between :math:`d_i` and :math:`q_i`,
+    and the paper reports results when using other such gating functions like sum or
+    concatenation.
+
+    Inputs:
+        - ``document_``, a matrix of shape ``(batch, document length, biGRU hidden length)``.
+          Represents the document as encoded by the biGRU.
+        - ``question_matrix``, a matrix of shape ``(batch, question length, biGRU hidden length)``.
+          Represents the question as encoded by the biGRU.
+        - ``normalized_qd_attention``, the soft attention over the document and question.
+          Matrix of shape ``(batch, document length, question length)``.
+
+    Output:
+        - ``X``, a tensor of shape ``(batch, document length, biGRU hidden length)`` if the
+          gating function is ``*`` or ``+``, or ``(batch, document length, biGRU hidden length * 2)``
+          if the gating function is ``||``  This serves as a representation of each token in
+          the document.
+
+    Parameters
+    ----------
+    gating_function : string, default="*"
+        The gating function to use for modeling the interactions between the document and
+        query token. Supported gating functions are ``"*"`` for elementwise multiplication,
+        ``"+"`` for elementwise addition, and ``"||"`` for concatenation.
+
+    Notes
+    -----
+    To find out how we calculated equation 1, see the GatedAttentionReader model (roughly,
+    a ``masked_batch_dot`` and a ``masked_softmax``)
+    """
+    def __init__(self, **kwargs):
+        self.supports_masking = True
+        # We need to wait until below to actually handle this, because self.name gets set in
+        # super.__init__.
+        # allowed gating functions are "*" (multiply), "+" (sum), and "||" (concatenate)
+        self.gating_function = kwargs.pop('gating_function', "*")
+        if self.gating_function not in GATING_FUNCTIONS:
+            raise ConfigurationError("Invalid gating function "
+                                     "{}, expected one of {}".format(self.gating_function,
+                                                                     GATING_FUNCTIONS))
+
+        super(GatedAttention, self).__init__(**kwargs)
+
+    def compute_mask(self, inputs, mask=None):
+        # pylint: disable=unused-argument
+        # We do not need a mask beyond this layer.
+        return None
```

Comment:
I second what Mark said.  In addition, I think you need to return a mask here, especially seeing how you use this in the model.  After the first run of the loop, you'll no longer have a mask.